NRF52 SDK15 update

TacoGrandeTX opened the pull request:
### Description

Add support for Nordic nRF5x SDK15.0

### Pull request type

    [ ] Fix
    [ ] Refactor
    [x] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

TacoGrandeTX committed:
Baseline legacy build using 14.2 drivers.

TacoGrandeTX committed:
Mods for baseline 15.0 support

TacoGrandeTX committed:
Bringing in Softdevice but excluding BLE feature and improvements in feature defines

TacoGrandeTX committed:
First porting effort covered by JIRA IOTDEV-1572:
- top level files ported from TARGET_NORDIC/TARGET_NRF5x/

Also addressed:
- fixed linking issue for gcc
- added support for nRF52-DK builds, but reverted to using nRF52840 sdk_config.h (must be updated)
- introduced "RTC" to targets.json (might need to be removed eventually)

TacoGrandeTX committed:
Ported analogin and pwm_out for IOTDEV-1573.

Also addressed:
- removed dependency on legacy config (excluded apply_old_config.h)
- removed legacy pwm and saadc headers
- Arm Compiler 5 linking issue (a band-aid for now... needs to
  be properly addressed for peripheral sharing)
- added missing header in SoftDevice file

Commit:
NRF52: i2c_api: port from Softdevice 14.2 to 15.0

Commit:
NRF52: spi_api: port from Softdevice 14.2 to 15.0

Commit:
serial_api: add changes to port from Softdevice 14.2 to 15.0

Commit:
serial_api: Fixes to compile with SDK15.0

TacoGrandeTX committed:
Updates to PWM for IOTDEV-1573

- Missed some NRFX defines that needed to changed
- Set PWM base clock to 125kHz (needs to be reverted back to 1 MHz)
- Updated sdk_config.h for nRF52_DK builds
- Brought in updates from PR7779 (fix for nRF52 PWM issues)

Commit:
Nordic sdk_config: Configure UART instances

Enable UART interfaces to enable serial debug prints.

Commit:
serial_api: Fix the interrupt registration for Nordic Serial Device

Using the nrfx_get_irq_number only works with the handle. Since we
know the IRQ numbers for UART0, RTC2 and EGU0, use them directly.

TacoGrandeTX committed:
Updates for IOTDEV-1575 and nRF52_DK builds
- Use new NRFX header file
- Fix nRF52832 linker script/ld files (hardcoded addresses for now)
- Temporarily remove DEVICE_TRNG for nRF52832 (which broke the build)
- Improve serial_putc() fix so we don't rely on "extra" functions

Commit:
NRF52: Fix vector table

Ensure that vector table gets initialized properly. The table that we
initialize in startup_nrf52840.S gets wiped out as the section is
declared as noinit. Fix this by implementing the weak function mbed_sdk_init
that inits the vector table.

Commit:
NRF52832: enable UART

Commit:
NRF: enable PRS (Peripheral Resource Sharing) for NRF52832 and NRF52840

TacoGrandeTX committed:
Updates for IOTDEV-1576 (only TRNG API)
- Added legacy nrf_drv_rng.c as there is no merit in fully adopting nrfx_rng.c
- Added nrf_queue library component
- Removed apply_old_config.h (unused for some time now)
- Updated sdk_config.h for queue and RNG support for nRF52832
- Brought back RNG into targets.json for nRF52832
Note: nRF52840 still uses CryptoCell 310 for TRNG

TacoGrandeTX committed:
nRF52832 needs NRFX_RNG_ENABLED to use the legacy driver

TacoGrandeTX committed:
nRF52_DK: Use mbed_lib.json to set app_offset (MBED_APP_START)

When the SoftDevice (SD) is enabled we need to set the app_offset
to 0x26000 to make room for the SoftDevice.  If we let the build
tools self-manage this, MBED_APP_START is coming out at 0x25000
which is not correct for the Nordic 15.0 SDK.

The app_offset value is translated to MBED_APP_START by the build
infrastructure.  We were hard coding MBED_APP_START in the scatter
and ld files as a temporary measure while testing.  This now sets
it properly if the SD is being brought in.

TacoGrandeTX committed:
Updates after rebase with Mbed master (circa OS v5.10.1)

* Remove NRF_SDH_CLOCK parameters from mbed_lib.json
* Bring in QSPI for nRF52840
* Migrate legacy QSPI driver to SDK v15 (nrfx_qspi.h)
* Remove outdated comment in i2c_api.c

TacoGrandeTX committed:
Fix build for nRF51 (fewer common files with nRF52)

The 15.0 SDK doesn't support the nRF51 so it must continue to use the legacy
NRF drivers.  Thus the original common_rtc.c, gpio_api.c and us_ticker.c are
restored and placed under TARGET_NRF5x/TARGET_NRF51.

Likewise the modified common_rtc.c, gpio_api.c and us_ticker.c are moved to
TARGET_NRF5x/TARGET_NRF52 so they are excluded from nRF51 builds.

TacoGrandeTX committed:
Remove dead #define PORT_SHIFT

nRF5x PinNames.h never utilized PORT_SHIFT so removed for clarity after
user commented on it.

Commit:
Add missing SoftDevice build support (MBR, NONE, OTA)

- Add MBR, NONE and OTA SoftDevice build options for S132 and S140
- Add S112 SoftDevice (single build option)
- Some folder restructuring in TARGET_SOFTDEVICE_COMMON was required

TacoGrandeTX committed:
Revert "NRF52: Fix vector table"

This reverts commit 3d2fa535a71b7bad08c6a91a31d74f4492c8948c.

This was a breaking change for the "MBR" and "NONE" builds.
After testing it was also determined that copying the vector
table a second time wasn't required for the "FULL" build.

TacoGrandeTX committed:
Cleanup for feature release

* Update TARGET_NRF5x/README.md to improve "Changing SoftDevice" section
  and added section on NRF52840 CryptoCell310 Support
* Update the file list in TARGET_SDK_15_0/TARGET_SOFTDEVICE_COMMON/README.md
* Add missing CR-LF to Nordic-provided SDK file
* Rename a header file in the TARGET_SOFTDEVICE_S112 tree

TacoGrandeTX committed:
Enabling NRFX_PPI_ENABLED for serial_api updates

TacoGrandeTX committed:
Changes required for rebase:
 - Brought in new nrfx APIs
 - Brought in PPI additions
 - Removed dead code for RTC

TacoGrandeTX committed:
Set the PWM clock prescaler back to default (1MHz clock)

Commit:
NRF52: serial: fix build warnings

Commit:
NRF52: Cleanup unused code

pan- commented:
@cmonr We can look at BLE changes included in this PR but I think you should require a review from the Hal team as most changes are in the HAL.

desmond-blue committed:
Update BLE SoftDevice from SDK14 to SDK15

    - Adjust memory for SoftDevice
    - Enable PRIO=5 for interrupt priority check
    - Change NRF_SD_BLE_API_VERSION to 6
    - Add handle and buffer for advertising and scanning
    - Remove guard for phy update
    - Change scatter files and mbed_lib.json for PR #8607

cmonr commented:
@JanneKiiskila Would you and your team like to review this as well?
This isn't coming into `master` yet, but figured I'vd give y'all the change to start looking at what's here now.

desmond-blue committed:
Fix timer1 irq configuration

0xc0170 commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_SDK_15_0/TARGET_SOFTDEVICE_S112/headers/ble_types.h:
```diff
@@ -0,0 +1,215 @@
+ * 5. Any software provided in binary form under this license must not be reverse
+ *    engineered, decompiled, modified and/or disassembled.
+ *
+ * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS
```
Not considering what is already in the codebase from previous SDKs, @TacoGrandeTX I'll email you shortly to review this license with licensing team.

elm3 replied:
Any word from the licensing team?  We are really hot to get this integrated.

TacoGrandeTX replied:
@elm3 No word from legal/licensing yet.

TacoGrandeTX replied:
@0xc0170 I think we are all satisfied with the off-line email exchange with Legal. Do you agree so I can resolve this?

0xc0170 replied:
I believe yes, we can focus now on the other parts.

0xc0170 commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/common_rtc.h:
```diff
@@ -0,0 +1,53 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2015 ARM Limited
```
add SPDX identifier and fix the year (new files 2018), also the company name

```
 * Copyright (c) 2018, Arm Limited and affiliates.
 * SPDX-License-Identifier: Apache-2.0
```

TacoGrandeTX replied:
@0xc0170 Is the following commit appropriate:  bc4aed00cb6c6ef66f566518e4cba164e0c5e298 ?  I've excluded any old files we didn't touch such as itm_api.c.  Would you want the company name and SPDX added for those files (and not touch the date)?   I'll then follow this model for any other Arm source files located under the TARGET_NORDIC tree.

0xc0170 replied:
Please follow for any, would be good 👍 

Thanks

desmond-blue committed:
Fix SoftDevice configuration on NRF52_DK

bulislaw commented on targets/targets.json:
```diff
@@ -6644,6 +6645,7 @@
             "PORTINOUT",
             "PORTOUT",
             "PWMOUT",
+            "RTC",
```
Are you sure these platforms actually support RTC as defined by Mbed? I remember RTC in some Nordic targets couldn't survive reset.

TacoGrandeTX replied:
Good question! @desmond-blue  - Can you comment?  Not sure why we have modified that Delta NRF51 target and not sure why we need the RTC for the NRF52840_DK.  Master doesn't have it defined and in sdk_config.h we are still not enabling RTC.

TacoGrandeTX replied:
Chatted with @desmond-blue.  This is a test left-over by me... I've corrected it with cfd4cb260ec284de45855886f1ddc61d9a841e0e.

bulislaw commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/TARGET_MCU_NRF52832/device/TOOLCHAIN_ARM_STD/nRF52832.sct:
```diff
@@ -9,17 +9,6 @@
   #define MBED_APP_SIZE 0x80000
 #endif
 
-/* If softdevice is present, set aside space for it */
```
What has changed that we don't need to block space for softdevice anymore?

TacoGrandeTX replied:
The new memory stats collection (specifically https://github.com/ARMmbed/mbed-os/pull/8607) broke Nordic SD builds as it defines MBED_RAM_START so the scatterfile needed to change.  We now define MBED_RAM_START and MBED_RAM_SIZE in the library json file when building with the SD:

https://github.com/TacoGrandeTX/mbed-os/blob/chile/targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_SDK_15_0/TARGET_SOFTDEVICE_S132_FULL/mbed_lib.json#L37-L39

This works out good because if one changes the BLE config the RAM requirements might also change and then they only need to touch one file.

desmond-blue replied:
We have only done on NRF52_DK and NRF52840_DK for now, I think that means we should add NRF51_DK to the verification, thank you for comments on this.

TacoGrandeTX replied:
@desmond-blue SDK15 doesn't include support for the NRF51 so there is no need to verify on the NRF51_DK.

bulislaw reviewed (commented):
What testing did you perform? As a minimum I would imagine all the tests on all base nrf boards + BLE testing should be performed locally.

0Grit commented:
Any chance of TWIM and SPIM support?

TacoGrandeTX committed:
Remove duplicate file

TacoGrandeTX committed:
Remove RTC from nRF52840_DK and DELTA_DFCM_NNN50 (was testing)

TacoGrandeTX committed:
Update file header at TARGET_NRF52 level
 - Update (c) year and company name
 - Add SPDX-License-Identifier
 - Impacted files:
    - Files touched for SDK15 port
    - Files added in 2018

pan- reviewed (dismissed):
So, I wanted to try and test ble but I faced several issues with that PR: 
- Serial reception doesn't work with this PR in our test application; looks like the RX interrupt is not triggered. 
- The following files should enable the BLE feature (reference [here](https://github.com/ARMmbed/mbed-os/blob/1c201b4628e346c62e1fbfa7cbcf6aa4e164f0c5/targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_SDK_14_2/TARGET_SOFTDEVICE_S132_FULL/mbed_lib.json#L36)): 
```
targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_SDK_15_0/TARGET_SOFTDEVICE_S132_FULL/mbed_lib.json
targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_SDK_15_0/TARGET_SOFTDEVICE_S132_OTA/mbed_lib.json
targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_SDK_15_0/TARGET_SOFTDEVICE_S140_FULL/mbed_lib.json   targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_SDK_15_0/TARGET_SOFTDEVICE_S140_OTA/mbed_lib.json
```

desmond-blue committed:
Add BLE feature to mbed_lib.json of SoftDevice

elm3 commented:
Hello Vincent,

Could you add the a link or reference to the test application where you see the issue?  We would really like to get this closed.

Thanks,
Earl

pan- commented:
Here is a small snippet that reproduces the issue: 

```c++
#include <stdio.h>

int main(void)
{
    while (true) {
        int c = getc(stdin);
        putc(c, stdout);
    }
    return 0;
}
```

@donatieng Maybe it would be useful to have a test as simple as this one to check that UART works. 

My own test case was a bit mot complex and involved IRQ and custom RawSerial; you can access it here: https://gist.github.com/pan-/9c836b51116aa8e9c3c262dca60c17b5 .

Commit:
NRF52: reloc_vector_table: check the RTS pin settings in code

Move the check from preprocessor to the if condition

naveenkaje commented:
Addressed the serial issue with https://github.com/TacoGrandeTX/mbed-os/pull/6

TacoGrandeTX commented:
Performed required rebase after test infrastructure broke resulting in a test failure for commit https://github.com/ARMmbed/mbed-os/commit/a770d038d8f400ca0c8d3b6e20af320622e0529c.

Greentea results are all pass for nRF52_DK and nRF52840_DK with the the exception of mbed_hal-trng.  This test also fails for my host system on master... the test performs a physical reset that doesn't play well with my laptop (results in a timeout).  Test results are attached for builds with the Arm compiler.

[greentea_log-arm_compiler.zip](https://github.com/ARMmbed/mbed-os/files/2785179/greentea_log-arm_compiler.zip)

cmonr commented:
@TacoGrandeTX Looking at the feature branch itself, since if this PR doesn't modify/update .travis.yml, it and any PRs opened against the feature branch will be hit with the same issues that Mbed OS faced last week.

_Fortunately_ since the feature-branch doesn't have any unique commits, rebasing it was not difficult.

This will need another rebase (sorry), but after that, it should be back to normal, testing-wise.

TacoGrandeTX commented:
@cmonr OK - thanks for that.  New rebase has been performed and testing looks good as before (attaching greentea reports).  The serial issue has been resolved so now @pan- can continue his review.

[gt_log-arm_compiler.zip.zip](https://github.com/ARMmbed/mbed-os/files/2789568/gt_log-arm_compiler.zip.zip)

desmond-blue committed:
Fix nrf section iter macro check

TacoGrandeTX committed:
Merge pull request #7 from desmond-blue/fix-nrf-section-iter

Fix nrf section iter ARMCC compiler macro check

cmonr commented:
@TacoGrandeTX Fyi: https://github.com/ARMmbed/mbed-os/pull/9358

0xc0170 commented:
@elm3 Please if PR needs merging talk to @ARMmbed/mbed-os-maintainers , this should not have been merged.

TacoGrandeTX commented:
@ARMmbed/mbed-os-maintainers Yes we are waiting for this to be merged.  What is the gating item?

TacoGrandeTX commented:
@cmonr I'm not expecting any impact to the SDK update from https://github.com/ARMmbed/mbed-os/pull/9358.  On the driver side the SDK update is largely a change in driver structure (not much has changed in functionality or implementation).  Likewise the new SoftDevice won't be impacted either.
